File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/6110a86f4218962475ccd7a3475ce2eb.py
import sys
from collections import deque
from heapq import *
from math import *
from bisect import *
from itertools import *
from functools import *
from operator import *
from fractions import *
from decimal import *
from random import *
from time import *
from copy import *
from queue import *

def read():
    global h, n, m, k, a
    h, n, m, k = map(int, input().split())
    a = []
    for _ in range(n):
        x, y = map(int, input().split())
        a.append((x, y))

def add(x):
    global mn
    q = deque()
    for i in range(k):
        q.append(i)
    while q:
        v = q.popleft()
        if mn[v] + x < mn[(v + x) % k]:
            mn[(v + x) % k] = mn[v] + x
            q.append((v + x) % k)

def recalc():
    global cur
    cur.clear()
    for i in range(n):
        if not used[i] and mn[a[i][0] % k] <= a[i][0]:
            cur.add((a[i][1], -i))

def solve():
    global mn, used, cur
    mn = [h + 1] * k
    mn[1] = 1
    used = [False] * N
    recalc()
    for _ in range(m):
        t = int(input())
        if t == 1:
            x = int(input())
            add(x)
            recalc()
        elif t == 2:
            x, y = map(int, input().split())
            x -= 1
            if mn[a[x][0] % k] <= a[x][0]:
                cur.remove((a[x][1], -x))
            a[x] = (a[x][0], a[x][1] - y)
            if mn[a[x][0] % k] <= a[x][0]:
                cur.add((a[x][1], -x))
        elif cur:
            print(cur[0][0])
            used[-cur[0][1]] = True
            cur.remove(cur[0])
        else:
            print(0)

while True:
    try:
        read()
        solve()
    except EOFError:
        break

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/6110a86f4218962475ccd7a3475ce2eb.java
import java.util.*;
import java.io.*;

class Main {
    static final int N = 100 * 1000 + 13;
    static int n, m, k;
    static long h;
    static Pair[] a;
    static boolean[] used;
    static long[] mn;
    static TreeSet<Pair<Integer, Integer>> cur;

    static class Pair<A, B> {
        A x;
        B y;

        Pair(A x, B y) {
            this.x = x;
            this.y = y;
        }

        public String toString() {
            return "(" + x + ", " + y + ")";
        }
    }

    static void add(long x) {
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < k; i++) {
            q.add(i);
        }
        while (!q.isEmpty()) {
            int v = q.poll();
            if (mn[v] + x < mn[(v + (int) x) % k]) {
                mn[(v + (int) x) % k] = mn[v] + x;
                q.add((v + (int) x) % k);
            }
        }
    }

    static void recalc() {
        cur.clear();
        for (int i = 0; i < n; i++) {
            if (!used[i] && mn[(int) (a[i].x % k)] <= a[i].x) {
                cur.add(new Pair<>(a[i].y, -i));
            }
        }
    }

    static void solve() {
        for (int i = 0; i < k; i++) {
            mn[i] = h + 1;
        }
        mn[1] = 1;
        used = new boolean[N];
        recalc();
        for (int i = 0; i < m; i++) {
            int t = in.nextInt();
            if (t == 1) {
                long x = in.nextLong();
                add(x);
                recalc();
            } else if (t == 2) {
                int x = in.nextInt();
                int y = in.nextInt();
                x--;
                if (mn[(int) (a[x].x % k)] <= a[x].x) {
                    cur.remove(new Pair<>(a[x].y, -x));
                }
                a[x].y -= y;
                if (mn[(int) (a[x].x % k)] <= a[x].x) {
                    cur.add(new Pair<>(a[x].y, -x));
                }
            } else if (!cur.isEmpty()) {
                out.println(cur.first().x);
                used[-cur.first().y] = true;
                cur.remove(cur.first());
            } else {
                out.println(0);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        while (read(br)) {
            solve();
        }
        out.close();
    }

    static boolean read(BufferedReader br) throws IOException {
        String[] line = br.readLine().split(" ");
        if (line.length != 4) {
            return false;
        }
        h = Long.parseLong(line[0]);
        n = Integer.parseInt(line[1]);
        m = Integer.parseInt(line[2]);
        k = Integer.parseInt(line[3]);
        a = new Pair[n];
        for (int i = 0; i < n; i++) {
            line = br.readLine().split(" ");
            a[i] = new Pair<>(Long.parseLong(line[0]), Integer.parseInt(line[1]));
        }
        return true;
    }
}
